Feat: suppot TFGrid-KYC

0oM4R opened the pull request:
### Description

Add KYC client to grid client, dashboard, and types packages.


changes: 

- grid client:
   - add kyc client
   - add kyc client to grid client
   - check if kyc verified on create name and handle methods in twinddeployment 
   
- types:  add related error classes 

- dashboard:
   - add kyc verifier that uses iframe to connect to idenfy
   - add kyc store, init on profile, clear if no profile
   - prevent deployment if kyc is not verified, in view layout
 
### tested scenarios:
 - try to deploy without kyc verification all deployments should be blocked
 - check twin page to see the kyc status
 - use kyc client and check its methods
 - check error handling in dashboard, toast etc, you can manipulate the headers in kyc client
 - check if you can deploy without kyc in gridclient
 - try to enter the verification process then cancel 
 - try to have rejected kyc verification
 - try to verify the kyc with valid data using the dashboard
 - try to scan the qr code  and continue verification process on mobile 
- make sure that the kyc status got cleaned after logout/refresh

related issues: 

- #3519 

info issue: https://github.com/threefoldtech/info_grid/issues/666

screenshots: 
![Screenshot from 2024-10-23 12-21-41](https://github.com/user-attachments/assets/13c7730e-fd85-4c30-a79d-17eee4df90a0)
![Screenshot from 2024-10-23 12-21-33](https://github.com/user-attachments/assets/1d3ce4cb-d64d-4746-a779-99b248713230)
![image](https://github.com/user-attachments/assets/a3af7d72-be03-4353-b9ac-934010d9a12d)
![Screenshot from 2024-10-23 12-27-40](https://github.com/user-attachments/assets/dba0d698-43c8-4bb9-a701-010221cb8b75)
![Screenshot from 2024-10-23 12-27-19](https://github.com/user-attachments/assets/236dee09-18c8-4e3e-8228-6a4162928c62)
![image](https://github.com/user-attachments/assets/6cf3bbcb-ac41-4bdc-b43c-eb21fac00b3c)
![image](https://github.com/user-attachments/assets/b13e125b-6347-4842-b62e-2f8aeb8c4bd1)

0oM4R committed:
Feat: add kyc to profile storage

0oM4R committed:
Feat: WIP adding KYC UI

0oM4R committed:
Feat: suppot TFGrid-KYC in gridclient

AhmedHanafy725 commented on packages/grid_client/src/clients/kyc/client.ts:
```diff
@@ -0,0 +1,131 @@
+   */
+  constructor(
+    public apiDomain: string,
+    public address: string,
```
why do we need an address here? you can get it from the mnemonic

AhmedHanafy725 commented on packages/grid_client/src/clients/kyc/client.ts:
```diff
@@ -0,0 +1,131 @@
+   * @param message - The message to convert.
+   * @returns The hexadecimal representation of the message.
+   */
+  private stringToHex(message: string): string {
```
this should be in utils

AhmedHanafy725 commented on packages/grid_client/src/clients/kyc/client.ts:
```diff
@@ -0,0 +1,131 @@
+   * @param hex - The hexadecimal string to convert.
+   * @returns {Uint8Array} The byte array representation of the hexadecimal string.
+   */
+  private bytesFromHex(hex: string): Uint8Array {
```
this one too should be in utils

AhmedHanafy725 commented on packages/grid_client/src/clients/kyc/client.ts:
```diff
@@ -0,0 +1,131 @@
+   */
+  async data(): Promise<VerificationDataResponse> {
+    const headers = await this.prepareHeaders();
+    return (await send("GET", `${this.apiDomain}/data`, "", headers)) as VerificationDataResponse;
```
```suggestion
    return (await send("GET", `${this.apiDomain}/api/v1/data`, "", headers)) as VerificationDataResponse;
```

AhmedHanafy725 replied:
please use url join instead

AhmedHanafy725 replied:
please apply this to the rest APIs

AhmedHanafy725 commented on packages/grid_client/src/clients/kyc/client.ts:
```diff
@@ -0,0 +1,131 @@
+   * @returns {Promise<Record<string, string>>} A promise that resolves to an object containing the prepared headers.
+   *
+   */
+  private async prepareHeaders(): Promise<Record<string, string>> {
```
why don't we use `KycHeaders` as the return type?

0oM4R replied:
the send method accepts `Record<string,string>` I'm using header type here to make it understandable only

0oM4R committed:
chore: export KYC client

0oM4R committed:
Chore:
- add stringToHex and bytesFromHex to utils
- remove private utils from KYC client and use them from utils

0oM4R committed:
chore : use urlJoin

0oM4R committed:
refactor: retrive address from mnemoinic

0oM4R committed:
chore: add api prefix as const

0oM4R committed:
Merge branch 'development' of github.com:threefoldtech/tfgrid-sdk-ts into development_kyc_gridClient

0oM4R committed:
Merge branch 'development_kyc' of github.com:threefoldtech/tfgrid-sdk-ts into development_kyc_gridClient

0oM4R committed:
prepare KYC client

0oM4R committed:
WIP: itegrate the kyc client in dashboard

0oM4R committed:
chore: update status response

0oM4R committed:
Feat: handle idenfy iframe

0oM4R committed:
chore: create kyc store

0oM4R committed:
chore: use kyc in env

0oM4R committed:
chore: use kyc store

0oM4R committed:
WIP: kyc style

0oM4R committed:
Style: enhance kyc dialog loading

0oM4R committed:
chore: add kyc to gridclient

0oM4R committed:
chore pass kyc url in gridclient options

0oM4R committed:
chore: handle domain post/prefix

0oM4R committed:
WIP: add kyc errors

0oM4R committed:
kyc: add token errors

0oM4R committed:
finalize errors

0oM4R committed:
add kyc client to twindeployment handler

0oM4R committed:
apply error handling

0oM4R committed:
add requirekyc

0oM4R committed:
update tooltip

0oM4R committed:
Merge branch 'development' of github.com:threefoldtech/tfgrid-sdk-ts into development_kyc_gridClient

0oM4R committed:
cleanup

0oM4R committed:
cleanup

0oM4R committed:
cleanup

0oM4R committed:
style: add icon, remove check

0oM4R committed:
UX: enhance token error handling

amiraabouhadid commented on packages/playground/src/utils/helpers.ts:
```diff
@@ -100,3 +100,25 @@ export function markAsFromAnotherClient(deployment: any): any {
+      case error instanceof KycErrors.KycInvalidAddressError:
+        return `${message}: Invalid address, please contact support`;
+      case error instanceof KycErrors.KycInvalidChallengeError:
+        return `${message}: Invalid challenge, please contact support`;
```
I don't think user would understand what 'invalid challenge' means. Can we rephrase this to perhaps 'Invalid Credentials' or 'Verification Failed'?

0oM4R replied:
As I discussed with @AhmedHanafy725, those parts of the message is for us as developers and support, we need to know what happen exactly, so we mentioned to contact support

amiraabouhadid commented on packages/grid_client/src/client.ts:
```diff
@@ -245,6 +247,8 @@ class GridClient {
       substrate: substrateURL || `wss://tfchain.${base}/ws`,
       graphql: graphqlURL || `https://graphql.${base}/graphql`,
       activation: activationURL || `https://activation.${base}/activation/activate`,
+      //TODO handle kyc url
```
remove this comment

0oM4R replied:
Needed to be handled later, on supporting kyc stacks

amiraabouhadid commented on packages/grid_client/src/clients/kyc/client.ts:
```diff
@@ -0,0 +1,190 @@
+      "X-Challenge": challenge,
+      "X-Signature": signedMsgHex,
+    };
+    return headers as unknown as Record<string, string>;
```
unknown?

0oM4R replied:
Yes we have to convert the type to record so we have to use it [ref](https://stackoverflow.com/questions/69399211/typescript-why-does-as-unknown-as-x-work)
What do you suggest?

amiraabouhadid commented on packages/grid_client/src/clients/kyc/client.ts:
```diff
@@ -0,0 +1,190 @@
+    switch (true) {
+      case errorMessage.includes("malformed address"):
+        throw new KycErrors.KycInvalidAddressError(errorMessage);
+      case errorMessage.includes("malformed challenge") || errorMessage.includes("bad challenge"):
```
since malformed is in all of these error messages, then why not just check for .includes("address"), .includes("challenge"), etc?

0oM4R replied:
the idea is those errors for handling corrupted headers. service may have another error that contains `address` and `challenge` so we need to make sure about the error message context.

amiraabouhadid commented on packages/grid_client/src/clients/kyc/client.ts:
```diff
@@ -0,0 +1,190 @@
+      const statusCode = (error as RequestError).statusCode;
+      this.ThrowHeadersRelatedError(errorMessage);
+      switch (true) {
+        case statusCode === 429:
```
put these statuscodes into an enum

amiraabouhadid commented on packages/grid_client/src/high_level/twinDeploymentHandler.ts:
```diff
@@ -34,6 +36,8 @@ class TwinDeploymentHandler {
     }
     try {
+      const kycStatus = await this.kyc.status();
+      if (kycStatus !== KycStatus.verified) throw new ValidationError("KYC is not verified.");
```
might be better as "Verification Failed", user doesn't necessarily know what KYC is.

samaradel reviewed (commented):
[Screencast from 23-10-24 15:34:39.webm](https://github.com/user-attachments/assets/902c34fe-7130-4072-9167-3424000b780c)
it keeps loading

0oM4R commented:
> [Screencast from 23-10-24 15:34:39.webm](https://github.com/user-attachments/assets/902c34fe-7130-4072-9167-3424000b780c) it keeps loading

did you build the clients ?

samaradel commented:
@0oM4R yes I did

samaradel commented:
> [Screencast from 23-10-24 15:34:39.webm](https://github.com/user-attachments/assets/902c34fe-7130-4072-9167-3424000b780c) it keeps loading

Update: It was from ```react extension``` was enabled and it works when it disabled.

0oM4R committed:
chore: use enum status code

0oM4R committed:
rephrase client error message

0oM4R committed:
fix typo

0oM4R committed:
chore: use new status endpoint
- cleanup code

0oM4R committed:
Fix:
- clear event listener before close kycdiloge
- get the info to kyc from config

0oM4R committed:
chore: enhance error messages

0oM4R committed:
Refactor:
enhance errors

0oM4R committed:
Refactor:
enhance errors

0oM4R committed:
fix: headers, enahnce error msg

0oM4R committed:
refactor:
enhance domain formatting

0oM4R committed:
WIP: handle diffrent networks

0oM4R committed:
docs: add kyc to build docs

xmonader commented on packages/grid_client/src/clients/kyc/client.ts:
```diff
@@ -0,0 +1,180 @@
+
+      case status === HttpStatusCode.PaymentRequired:
+        return new InsufficientBalanceError(errorMessage);
+    }
```
And else?

0oM4R replied:
default case was removed by mistake, added now

xmonader commented on packages/types/src/errors/grid_client/kyc.ts:
```diff
@@ -0,0 +1,53 @@
+  Unverified,
+  InvalidResponse,
+  RateLimit,
+  badRequest,
```
Why the inconsistency? Change to BadRequest, Unauthorized!

0oM4R replied:
updated; thank you!

xmonader commented on packages/grid_client/src/high_level/twinDeploymentHandler.ts:
```diff
@@ -502,6 +509,11 @@ class TwinDeploymentHandler {
+    const kycStatus = await this.kyc.status();
+    if (kycStatus !== KycStatus.verified)
+      throw new ValidationError(
+        "Your account is not verified. Please sign into Threefold Dashboard or Connect mobile app to complete your KYC verification.",
```
Revise the app name it's called `Threefold Connect`

0oM4R replied:
replaced with ` Your account is not verified. Please sign into ThreeFold Dashboard or ThreeFold Connect mobile app to complete your KYC verification.`

0oM4R committed:
refactor:
- add default case to throwkycError method
- make errors enum values consist with each other
- rephrase error message in twinDeploymentHandler

0oM4R committed:
feat: add isHealty to kyc client

AhmedHanafy725 commented on packages/grid_client/src/high_level/twinDeploymentHandler.ts:
```diff
@@ -34,6 +36,11 @@ class TwinDeploymentHandler {
       }
     }
     try {
+      const kycStatus = await this.kyc.status();
+      if (kycStatus !== KycStatus.verified)
+        throw new ValidationError(
+          "Your account is not verified. Please sign into ThreeFold Dashboard or ThreeFold Connect mobile app to complete your KYC verification.",
+        ); //TODO add kyc manual link
```
this is not required, already domains require 2 contracts (name contract to be unique and node contract to deploy it)

0oM4R committed:
chore: remove kyc check on createNameContract
